### PR TITLE
JSPM and SystemJS recipe.

### PR DESCRIPTION
**Pull request by skorlir**
A recipe for using AVA with JSPM, per discussion on #131. It requires the import of an external library I wrote for the purpose of encapsulating the loader shim so it can be maintained and updated. Here's a link to that library: [ava-jspm-loader](https://github.com/skorlir/ava-jspm-loader).

### CODE REVIEW

**Comment by dak** [comment]
This appears to work for normalizing mappings, but not for paths, although the `ava-jspm-loader` also appears to only focus on paths.

This may be because I'm only using JSPM to bundle dependencies and not the entire app.

I wonder if we may need a better synchronous API from JSPM to properly let it handle loading modules using its configs.

**Comment by skorlir** [comment]
@dak 
When you say, "not for paths", what do you mean?

Currently, if you have a System config mapping like:

```
System.config({
  browserConfig: {
    "paths": {
      "app/": "/src/"
    }
  }
})
```

And then you try to import `app/main.js`, this will resolve correctly to `/src/main.js`. That's typical SystemJS behaviour. If you aren't mapping your source files in your SystemJS config, it won't be able to find your source files that way. Of course, you could probably still do something like `import main from '../src/main'`, but that's less ideal.

Why are you only bundling part of your project using SystemJS?

Can you give a concrete example of a configuration that does not work correctly?

**Comment by dak** [comment]
I'm only bundling part of it because I'm hosting over HTTP/2 and the app is quite large, so the benefit of reducing the number of requests is near zero due to multiplexing (just the extra headers) compared with being able to return just the relevant templates/scripts for a given page.
It's a supported use case by systemjs builder: https://github.com/systemjs/builder#example---third-party-dependency-bundles

I'm also using JSPM 0.17, which may? be part of the discrepancy, although I suspect it's more than that.

Stripped down, my config is basically the same as above:
```js
paths: {
    "~/": "src/app/"
}
```
but I'll get errors like: `Error: Cannot find module '~/path/to/module'`, which suggest the path normalization never occurred.

Let me try to make a simple repo to demo the issue.

**Comment by skorlir** [comment]
[This part](https://github.com/skorlir/ava-jspm-loader/blob/master/jspm-loader.js#L16) of ava-jspm-loader should account for that path mapping, so it strikes me as odd that this isn't working. Let me know if you get a demo repo up!

**Comment by dak** [comment]
@skorlir See: https://github.com/dak/jspm-ava-test

**Comment by skorlir** [comment]
@dak I made a pull request into your example project that fixes the testing configuration.

Your project was just misconfigured. I've spent a bunch of time on JSPM and AVA configuration already this week, so I feel your pain. You have to get it just right.

**Comment by skorlir** [comment]
Actually, there is a bug there - you pointed it out but I didn't see it.

The jspm-ava-loader doesn't handle nested paths. Just a sec and I'll push a fix,

**Comment by skorlir** [comment]
It's fixed now and working correctly. The recipe is unchanged, but ava-jspm-loader has been patched to handle directory nesting.

**Comment by dak** [comment]
@skorlir I think that did the trick. :+1:

**Comment by MeoMix** [comment]
@skorlir When I try using your repo I encounter an error:

> Error: Cannot find module '/C:/Users/Meo/GitHub/jspm-ava-test/jspm_packages/github/jspm/nodelibs-fs@0.2.0-alpha'

None of the node-libs seem to be loading properly.

If I work around it by not applying `System.normalizeSync` to nodelibs then I encounter another error:

> Error: Cannot find module '/C:/Users/Meo/GitHub/jspm-ava-test/src/app/test/test.js'

Suspect you guys aren't testing this against Windows paths?

**Comment by dak** [comment]
@skorlir @MeoMix I just noticed something similar actually... you need to change the order and prefer NPM modules over JSPM modules. It broke when I tried combining it with nyc for the same reason @MeoMix identified.

Reversing the loading order (with a try/catch) resolved the issue.

**Comment by MeoMix** [comment]
@dak Yeah, that appears to work, too -- although using try/catch as part of normal control flow is generally frowned upon.

Additionally, there are still issues with Windows paths. I need to have the following:

`jspmUri = jspmUri.replace('///', '//');` for it to work properly as Windows looks like `file:///C:/...` which results in the prefix'ed slash which causes lookup failure.

**Comment by dak** [comment]
Rather than trying to duplicate how JSPM is loading modules, it would be nice if JSPM exposed one or two methods to handle this. /me nudges @guybedford
I'm sure we're missing other edge cases still.

**Comment by MeoMix** [comment]
Well, I'm pretty sure that's what JSPM does internally. It's a bit of an odd scenario to interface into JSPM, but then go back to loading through NPM after using it.

I have to wonder if we can just do something similar to how I've handled it with the JSPM css loader: https://github.com/MeoMix/jspm-loader-css/blob/master/src/abstractLoader.js#L58

but I'm not sure atm if they just seem similar or if it's actually the same problem being solved

**Comment by skorlir** [comment]
The Node errors are interesting. I don't get those... I will have to look into it further. You are probably right that we should just try to load from node_modules first, but I don't like it. I also suspect this will break module loading error messages if we are not careful. 

I can fix the Windows paths, but I am confused why a Windows path would even matter? If it isn't a path in System.map (which those paths are platform agnostic) the function in question should return false. I think any absolute path, Windows or otherwise, will return false - which was the behavior I expected. It could very well be the case that it is possible to have a leading slash in a map path, but that strikes me as unlikely since everything is baseUrl relative.

I agree however: that function is an area for improvement. The nested paths patch made it messy and it is likely there are more edge cases there. 

@dak I think the intent is that you'll just use jspm.import (which itself is just an interface into SystemJS). Unfortunately there's no synchronous import - that's what I would have wanted to use really. I tried to think of a way to "synchify" it but it made my head hurt. 

@meomix I can't use the exact fix you are using for the css loader partly because I can't do any async work in the Module._load shim. I am not sure I should manually prepend baseUrl either - I'd rather get normalization working. Was there something else there you wanted me to look at?

**Comment by skorlir** [comment]
@dak can you give a sample setup where `nyc` or some other node module breaks?

**Comment by dak** [comment]
@skorlir See: https://github.com/dak/jspm-ava-test again
You'll need nyc installed globally, I didn't add it to the package (or npm install it manually). The run `npm test`.

**Comment by dak** [comment]
@skorlir So I think I've found an even better solution by making use of Ava's support for async functions.  We kept trying to get around how JSPM loads modules asynchronously instead of just embracing it.

The `ava-jspm-loader` would simply be:
```js
const path = require('path');
const jspm = require('jspm');

const pjsonLoc = require('find-pkg').sync(process.cwd());
const pjsonDir = path.parse(pjsonLoc).dir;

jspm.setPackagePath(pjsonDir);

global.System = jspm.Loader();
```

No funky hacks.

Then, you'd write tests using async functions:

import test from 'ava';

test('Test name', async function (assert) {
    const module = await System.import('jspm/import/path');

    let actual = module.returnTrue();
    let expected = true;

    assert.is(actual, expected, 'returnTrue() should return true.');
});
```

**Comment by skorlir** [comment]
Interesting - you must be bundling your code before testing if that works
for you?

**Comment by dak** [comment]
@skorlir It does appear to need to run through Babel first... would it work with babel-register if you included the systemjs plugin for babel too?

**Comment by skorlir** [comment]
Can you test modules in your source code separately this way? Can you test
"internal" modules that aren't exposed as part of the API?

If you have to bundle first, I'm guessing not.

If so, though, this would be worth looking into further.

**Comment by dak** [comment]
Yes, I build them but don't bundle them, so they're still individual files.
See: https://github.com/dak/jspm-ava-test
After you run `npm test` take a look at the `dev` folder it creates.

File: docs/recipes/jspm-systemjs.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# JSPM + SystemJS for ES6
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: ES6 => ES2015

File: docs/recipes/jspm-systemjs.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+### Babel
+
+Configure your .babelrc to work with AVA if you have not already. NOTE: You can keep
+additional configuration in your JSPM config files to override these settings during
+bundling and building.
+
+```json
+{
+  "presets": [ "es2015", "stage-2" ]
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]:
```json
"presets": ["es2015", "stage-2"]
```

File: docs/recipes/jspm-systemjs.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+You can find more information about setting up Babel with AVA in the [babelrc recipe](babelrc.md).
+
+### JSPM Loader Helper
+
+You will need to install the AVA JSPM Loader as a dev dependency.
+
+`npm install --save-dev ava-jspm-loader`
+
+You will also need to update your AVA config in package.json to use the jspm
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: jspm => JSPM

File: docs/recipes/jspm-systemjs.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+{
+  "presets": [ "es2015", "stage-2" ]
+}
+```
+
+You can find more information about setting up Babel with AVA in the [babelrc recipe](babelrc.md).
+
+### JSPM Loader Helper
+
+You will need to install the AVA JSPM Loader as a dev dependency.
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: Loader => loader

File: docs/recipes/jspm-systemjs.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    "require": [
+      "babel-register",
+      "ava-jspm-loader"
+    ]
+  }
+}
+```
+
+NOTE: if you use async/await in your source code (not in your test code), you
+will need to install "babel-polyfill" from NPM and add it to your require array.
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: NPM => npm

File: docs/recipes/jspm-systemjs.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  "ava": {
+    "require": [
+      "babel-register",
+      "ava-jspm-loader"
+    ]
+  }
+}
+```
+
+NOTE: if you use async/await in your source code (not in your test code), you
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: if => If

File: docs/recipes/jspm-systemjs.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+### Example test file
+
+Note that you will need to use `System.import` paths for all of your project
+files. So, if you named your project `app` and you want to import your `main.js`
+into a test file, you will need to `import main from 'app/main'`.
+
+```js
+import test from 'ava'
+
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: No need for empty line here.

File: docs/recipes/jspm-systemjs.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import main      from 'app/main'     // maps to your JSPM config for "app/main.js"
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: Don't align the `from`.

File: docs/recipes/jspm-systemjs.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import BigNumber from 'bignumber.js' // in jspm_packages
+
+function fn () {
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]:
```js
function fn() {
```

File: docs/recipes/jspm-systemjs.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  return Promise.resolve(new BigNumber("1234567890.123456789"))
+}
+
+test('example test', async t => {
+  t.is((await fn()).toString(), "1234567890.123456789")
+})
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: single-quotes, tab indent, and semicolons

applies to all the code examples

**Comment by sindresorhus** [comment]
Can you add it to the reamde: https://github.com/avajs/ava#recipes

File: docs/recipes/jspm-systemjs.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
+# JSPM + SystemJS for ES6
+
+It requires a special loader helper to correctly resolve `import`s of JSPM
```
Comment (sindresorhus) [inline docs/recipes/jspm-systemjs.md]: Packages => packages

**Comment by skorlir** [comment]
@dak released a patch of ava-jspm-loader today that was (tangentially) related to `nyc` and may actually apply to your earlier issue

@sindresorhus is there anything else I should do for this PR? I made your requested changes.

**Comment by kdex** [comment]
@skorlir @dak [sync](https://www.npmjs.com/package/sync) might be a way to wrap `jspm.import` in a synchronous way so people could import their jspm dependencies using the `import` syntax.

**Comment by sindresorhus** [comment]
Looks good! :)

Thanks @skorlir and @dak.

**Pull request closed without merging**